[mlir][emitc] Fix creating pointer from constant array

Jimmy2027 opened the pull request:
When creating a pointer from a constant emitc array, check if it is constant. If it is, create the pointer as opaque<"const {type}">>.

Move out C type string creation logic from TranslateToCpp.cpp to getCTypeString in EmitC.cpp as a shared utility function.

Jimmy2027 commented:
> But you can't correctly use pointer to const int right? In mlir you can pass the result of get_global from a constant global in a function call. In C you wouldn't be allowed to pass this to a function expecting a ptr to int/int array.
> 
> So I think you will get compile errors in the generated code dependent in the users of the get_global with your changes right now. But maybe I'm wrong, I'd have to test this first.

But wouldn't that be a user mistake in any case? C++ doesn't even compile a non-const pointer of a const... I can't think of any meaningful example where this change breaks something that is not already broken but maybe I'm biased and missing something :)
The use case I'm working with is the one from the [testcase I added](https://github.com/llvm/llvm-project/pull/162083/files#diff-64724a9af240143a2f133042d3c016c9486d38588712c3f4e7d44ad622f6e6adR78) where I memcpy the const data somewhere. Without the fix this would currently not compile with g++

RAGL1701 commented:
> When creating a pointer from a constant emitc array, check if it is constant. If it is, create the pointer as opaque<"const {type}">>.
> 
> Move out C type string creation logic from TranslateToCpp.cpp to getCTypeString in EmitC.cpp as a shared utility function.

Jimmy2027 commented:
> This doesn't seem like an emitc-specific issue (i.e. any dialect wishing to model that property in their type system would have a problem doing local type conversions) so this might be worth a wider discussion, e.g. are memref's memory spaces the proper way to carry this information? should memrefs support the constant property directly?

If I understand correctly at the memref level, it is expected to model the constness with side effects and ops (memref.global constant). I'm not sure if something like a const attr to the memref type breaks its philosophy...

> At least in the short term I think the simplest solution is indeed to cast the constness away when lowering `get_global`. It actually reflects the semantic knowledge the memrefs carry (the array is constant at definition but we don't have such information at the users) and it should be fairly easy to implement.

Yeah maybe this is the safest way forward for this PR, though I still can't think of any meaningful example where casting away the constness is actually needed... 

> In the long run, adding a `const` attribute to emitc.ptr (or even to any emitc type) could be useful if there's any user that lowers to emitc a dialect that actually maintains the constant property in its types. We could also try to deduce it for memrefs by writing some global analysis to decide which pointer is constant. For example, since C allows passing non-constant where a constant is required we could try to mark pointers by default as constant and only mark pointers as non-constant if they reach `emitc.assign` ops (arbitrary C code might be too difficult to analyze but scf+memref-based code would perhaps be simple enough).

yup I agree, that would be the nicest solution. The case I'm currently working with is actually obvious since the memref comes from a const global

Jimmy2027 committed:
[mlir][emitc] Emit casting away constness when taking address of const global

- Modify the C++ emitter to detect when an AddressOf op traces back to a const global. If it does, emit a
C-style cast (T*)(&...) to strip the const qualification
- Adapt mlir/test/Target/Cpp/global.mlir to check for correct syntax of
  the generated code

aniragil commented on mlir/lib/Target/Cpp/TranslateToCpp.cpp:
```diff
@@ -397,6 +397,51 @@ static bool shouldBeInlined(ExpressionOp expressionOp) {
+    return nullptr;
+
+  SymbolTable symbolTable(symbolTableOp);
+  auto globalOp = symbolTable.lookup<emitc::GlobalOp>(getGlobalOp.getName());
```
Could use `SymbolTable::lookupNearestSymbolFrom()`?

Jimmy2027 committed:
use lookupNearestSymbolFrom

aniragil commented on mlir/test/Target/Cpp/global.mlir:
```diff
@@ -1,5 +1,11 @@
 // RUN: mlir-translate -mlir-to-cpp %s | FileCheck %s -check-prefix=CPP-DEFAULT
 // RUN: mlir-translate -mlir-to-cpp -declare-variables-at-top %s | FileCheck %s -check-prefix=CPP-DECLTOP
+// check that the generated code is syntactically valid
+// RUN: mlir-translate -mlir-to-cpp %s | %host_cxx -fsyntax-only -x c++ -
```
So there are some aspects here I think might be worth a wider discussion:

- Implementation: IIUC the `%host_cxx` is the compiler (and version) used for building LLVM itself. It won't necessarily align with EmitC's restrictions, and we'd also be adding to it a new use-case unknown to folks not involved in EmitC.

- Compile time: going through a C++ frontend may affect the total runtime of `check-mlir`.

- Scope: As in this case, making the result pass the frontend requires adding code which isn't part of the test (in this case include directives), which goes against the LLVM guideline to keep LIT tests minimal.

- Are LIT tests the right tool for this? AFAIK LIT tests don't usually make sure the expected result is "semantically" correct - in this case that the emitted code is valid C code - only that it is as expected (however the SPIRV translator does seem to do that: its LIT tests run SPIRV validation under a `%if spirv-tools` condition). Alternatives to LIT tests might be LLVM's e2e test-suite or setting up a dedicated buildbot.

Would be good to hear what others think about this, e.g. on discourse (and if such a mechanism is constructed, introduce it in its own PR consistently for all relevant tests). I'd leave it out of this PR, both to get the bug fix in asap and to avoid having the bug fix reverted if this kind of testing proves problematic. WDYT?

Jimmy2027 committed:
Remove syntax validity check

aniragil reviewed (approved):
LGTM, thanks!
